brainlit/algorithms/generate_fragments/spline_fxns.py

```diff
@@ -0,0 +1,296 @@
+import numpy as np
+from typing import Iterable, Union, Tuple
+from brainlit.utils.util import (
+    check_type,
+    check_size,
+    check_precomputed,
+    check_iterable_type,
+    check_iterable_nonnegative,
+)
+
+
+def speed(
+    x: np.ndarray,
+    t: np.ndarray,
+    c: np.ndarray,
+    k: np.integer,
+    aux_outputs: bool = False,
+) -> np.ndarray:
+    """Compute the speed of a B-Spline.
+
+    The speed is the norm of the first derivative of the B-Spline
+
+    Arguments:
+        x: A 1xL array of parameter values where to evaluate the curve
+        t: A 1xm array representing the knots of the B-spline
+        c: A dxn array representing the control points of the B-spline
+        k: An integer representing the degree of the B-spline:
+
+    Returns:
+        speed: A 1xL array containing the speed of the B-Spline evaluated at x
+    """
+
+    # convert arguments to desired type
+    x = np.ascontiguousarray(x)
+    t = np.ascontiguousarray(t)
+    c = np.ascontiguousarray(c)
+    k = operator.index(k)
+
+    if k < 0:
+        raise ValueError("The order of the spline must be non-negative")
+
+    check_type(t, np.ndarray)
+    t_dim = t.ndim
+    if t_dim != 1:
+        raise ValueError("t must be one-dimensional")
+    if len(t) == 0:
+        raise ValueError("t must be non-empty")
+    check_iterable_type(t, (np.integer, np.float))
+    if (np.diff(t) < 0).any():
+        raise ValueError("t must be a non-decreasing sequence")
+
+    check_type(c, np.ndarray)
+    c_dim = c.ndim
+    if c_dim > 2:
+        raise ValueError("c must be 2D max")
+    if len(c.flatten()) == 0:
+        raise ValueError("c must be non-empty")
+    if c_dim == 1:
+        check_iterable_type(c, (np.integer, np.float))
+        # expand dims so that we can cycle through a single dimension
+        c = np.expand_dims(c, axis=0)
+    if c_dim == 2:
+        for d in c:
+            check_iterable_type(d, (np.integer, np.float))
+    n_dim = len(c)
+
+    check_type(x, np.ndarray)
+    x_dim = x.ndim
+    if x_dim != 1:
+        raise ValueError("x must be one-dimensional")
+    if len(x) == 0:
+        raise ValueError("x must be non-empty")
+    check_iterable_type(x, (np.integer, np.float))
+    L = len(x)
+
+    # evaluate first and second derivatives
+    # deriv, dderiv are (d, L) arrays
+    deriv = np.empty((n_dim, L))
+    for i, dim in enumerate(c):
+        spl = BSpline(t, dim, k)
+        deriv[i, :] = spl.derivative(nu=1)(x) if k - 1 >= 0 else np.zeros(L)
+    # tranpose deriv
+    deriv = deriv.T
+
+    speed = np.linalg.norm(deriv, axis=1)
+    if aux_outputs == False:
+        return speed
+    else:
+        return speed, deriv
+
+
+def curvature(
+    x: np.ndarray,
+    t: np.ndarray,
+    c: np.ndarray,
+    k: np.integer,
+    aux_outputs: bool = False,
+) -> np.ndarray:
+    """Compute the curvature of a B-Spline.
+
+    The curvature measures the failure of a curve to be a straight line.
```

Comment:
Reference where you found the below equations

Reply:
I found them [here](https://en.wikipedia.org/wiki/Frenet%E2%80%93Serret_formulas#Other_expressions_of_the_frame), do you want me to reference them in the code too?